cpp/src/util/main/UnitTests.cpp

```diff
@@ -125,6 +125,343 @@ TEST(eigen_util, sort_columns) {
+
+  Array array{{0, 1, 2, 3}};
+  Array expected{{0.5, 0.7310586, 0.8807971, 0.9525741}};
+
+  array = eigen_util::sigmoid<Array>(array);
+
+  for (int i = 0; i < N; ++i) {
+    EXPECT_NEAR(array(i), expected(i), 1e-5);
+  }
+}
+
+TEST(eigen_util, reverse) {
+  constexpr int M = 3;
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<3, 4>>;
+
+  Tensor tensor;
+  tensor.setValues({{0, 1, 2, 3}, {4, 5, 6, 7}, {8, 9, 10, 11}});
+  Tensor expected;
+  expected.setValues({{3, 2, 1, 0}, {7, 6, 5, 4}, {11, 10, 9, 8}});
+
+  Tensor reversed = eigen_util::reverse<Tensor>(tensor, 1);
+
+  for (int i = 0; i < M; ++i) {
+    for (int j = 0; j < N; ++j) {
+      EXPECT_EQ(reversed(i, j), expected(i, j));
+    }
+  }
+}
+
+TEST(eigen_util, normalize) {
+  constexpr int M = 2;
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<M, N>>;
+  using Array = Eigen::Array<float, M, N, Eigen::RowMajor>;
+
+  Array array{{0, 1, 2, 3}, {0, 0, 4, 0}};
+  Array expected = array / array.sum();
+  Tensor tensor = Eigen::TensorMap<Tensor>(array.data(), array.rows(), array.cols());
+
+  bool success = eigen_util::normalize<Tensor>(tensor, 1e-5);
+
+  EXPECT_TRUE(success);
+  for (int i = 0; i < M; ++i) {
+    for (int j = 0; j < N; ++j) {
+      EXPECT_NEAR(tensor(i, j), expected(i, j), 1e-5);
+    }
+  }
+}
+
+TEST(eigen_util, sample) {
+  constexpr int N = 4;
+  constexpr int numSamples = 10000;
+
+  using Array = eigen_util::FArray<N>;
+
+  Array values = {{1, 2, 3, 4}};
+  Array expectedFreq = values / values.sum() * numSamples;
+
+  Array freq;
+  freq.setZero();
+  for (int i = 0; i < numSamples; i++) {
+    auto sample = eigen_util::sample<Array>(values);
+    freq(sample)++;
+  }
+
+  // Chi-Squared Statistic \sum{(O_i - E_i)^2 / E_i}
+  // where O_i is the observed frequency and E_i is the expected frequency
+  float chi2 = 0;
+  for (int i = 0; i < N; i++) {
+    float diff = freq(i) - expectedFreq(i);
+    chi2 += diff * diff / expectedFreq(i);
+  }
+
+  // The chi-squared value corresponding to a p-value of 0.05 with 3 degrees of freedom is 7.81
+  EXPECT_LT(chi2, 7.81);
+}
+
+TEST(eigen_util, randomly_zero_out) {
+  constexpr int M = 8;
+  constexpr int N = 4;
+  constexpr int numZeroes = 10;
+  static_assert(numZeroes <= M * N, "numZeroes must be less than or equal to M * N");
+
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<M, N>>;
+  Tensor tensor;
+  tensor.setRandom();
+
+  eigen_util::randomly_zero_out(tensor, numZeroes);
+
+  auto data = tensor.data();
+  auto size = tensor.size();
+  int zero_count = 0;
+
+  for (int i = 0; i < size; ++i) {
+    if (data[i] == 0) {
+      zero_count++;
+    }
+  }
+  EXPECT_GE(zero_count, 10);
+}
+
+TEST(eigen_util, reinterpret_as_array) {
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<N, 1>>;
+
+  Tensor tensor;
+  tensor.setValues({{0}, {1}, {2}, {3}});
+
+  auto array = eigen_util::reinterpret_as_array(tensor);
```

Comment:
This test would be a bit better if it directly constructed an `Array expected_array`, and then compared `array` against `expected_array`, rather than comparing `array` against `tensor`.